modules/foreman_user.py

```diff
@@ -0,0 +1,434 @@
+  mail:
+    description:
+      - Email address of the user
+      - Required when creating a new user
+    required: false
+  description:
+    description:
+      - Description of the user
+    required: false
+  admin:
+    description:
+      - Whether or not the user is an administrator
+    required: false
+    default: false
+    type: bool
+  user_password:
+    description:
+      - Password for the user
+    required: false
+  default_location:
+    description:
+      - The location that the user uses by default
+    required: false
+  default_organization:
+    description:
+      - The organizxation that the user uses by default
+    required: false
+  auth_source:
+    description:
+      - Authentication source where the user exists
+    required: false
+    default: Internal
+  timezone:
+    description:
+      - Timezone for the user
+      - If blank it will use the browser timezone.
+    required: false
+  locale:
+    description:
+      - The language locale for the user
+    required: false
+  roles:
+    description:
+      - List of roles assigned to the user
+    required: false
+    type: list
+  locations:
+    description:
+      - List of locations assigned to the user
+    required: false
+    type: list
+  organizations:
+    description:
+      - List of organizations assigned to the user
+    required: false
+    type: list
+  state:
+    description:
+      - State of the user
+    default: present
+    choices:
+      - present
+      - absent
+'''
+
+EXAMPLES = '''
+- name: Create a user
+  foreman_user:
+    name: test
+    firstname: Test
+    lastname: Userson
+    mail: test.userson@example.com
+    description: Dr. Test Userson
+    admin: no
+    user_password: s3cret
+    default_location: Test Location
+    default_organization: Test Organization
+    auth_source: Internal
+    timezone: Stockholm
+    locale: sv_SE
+    roles:
+      - Manager
+    locations:
+      - Test Location
+    organizations:
+      - Test Organization
+    state: present
+
+- name: Update a user
+  foreman_user:
+    name: test
+    firstname: Tester
+    state: present
+
+- name: Change password
+  foreman_user:
+    name: test
+    user_password: newp@ss
+
+'''
+
+RETURN = ''' # '''
+
+from ansible.module_utils.foreman_helper import (
+    sanitize_entity_dict,
+    ForemanEntityApypieAnsibleModule,
+)
+
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'login',
+    'firstname': 'firstname',
+    'lastname': 'lastname',
+    'mail': 'mail',
+    'description': 'description',
+    'admin': 'admin',
+    'user_password': 'password',
+    'default_location': 'default_location_id',
+    'default_organization': 'default_organization_id',
+    'auth_source': 'auth_source_id',
+    'timezone': 'timezone',
+    'locale': 'locale',
+    'roles': 'role_ids',
+    'locations': 'location_ids',
+    'organizations': 'organization_ids',
+}
+
+# List of allowed timezones
+timezone_list = [
```

Comment:
Can we pull these from apypie? It's should be exposed via https://github.com/Apipie/apypie/blob/master/apypie/param.py#L19

Reply:
That would be awesome. I just need to figure out how to do that.

Reply:
First of all, I don't understand Apypie well enough to know how to fetch that data from the server.

Secondly, as far as I can tell when checking `/apidoc/v2.json` is that the `validator` field just contains a human description and I'm not sure if that is stable enough for us to parse it and use it for validation of the module params.

Perhaps we could replace the hardcoded list in a later PR?

Reply:
I was just about to comment something similar. We're going to need changes to apypie. :+1:

Reply:
Can you please open an issue against apypie? The validators can use some love indeed ;)

Reply:
https://github.com/Apipie/apypie/issues/27